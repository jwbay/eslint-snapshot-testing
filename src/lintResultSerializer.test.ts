import { Linter, Rule } from 'eslint'
import { serializeLintResult } from './'
import { VariableDeclarator } from 'estree'

const processCwd = jest.spyOn(process, 'cwd')
const sourceCode = `\nconst foo = 'something';`

beforeEach(() => {
	processCwd.mockReturnValue('/cwd')
})

describe('serializeLintResult supports per-test setup and mocking for lint rules', () => {
	test('should not have errors', () => {
		processCwd.mockReturnValue('/foo')
		expect(lint(sourceCode)).toMatchSnapshot()
	})

	test('should have errors', () => {
		processCwd.mockReturnValue('/bar')
		expect(lint(sourceCode)).toMatchSnapshot()
	})
})

function lint(source: string) {
	const linter = new Linter({})
	linter.defineRule('my-rule-name', testRule)
	const lintMessages = linter.verify(source, {
		rules: { ['my-rule-name']: 'error' },
		parserOptions: {
			ecmaVersion: 2020,
		},
	})
	return serializeLintResult({
		lintedSource: source,
		lintMessages,
	})
}

const testRule: Rule.RuleModule = {
	create(context) {
		return {
			VariableDeclarator(node) {
				const variableDeclarator = node as VariableDeclarator

				if (
					variableDeclarator.id?.type === 'Identifier' &&
					variableDeclarator.id.name === 'foo' &&
					process.cwd().startsWith('/bar')
				) {
					context.report({
						loc: variableDeclarator.id.loc!,
						message:
							'variables named foo are not allowed when working directory is /bar',
					})
				}
			},
		}
	},
}
