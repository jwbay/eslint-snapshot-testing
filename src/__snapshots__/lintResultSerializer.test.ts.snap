// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeLintResult supports per-test setup and mocking for lint rules should have errors 1`] = `
"
const foo = 'something';
      ~~~~~~~~~~~~~~~~~  [1]
[1] variables named foo are not allowed when working directory is /bar"
`;

exports[`serializeLintResult supports per-test setup and mocking for lint rules should not have errors 1`] = `
"
const foo = 'something';"
`;
